src/Microsoft.AspNet.OData.Shared/Formatter/ClrTypeCache.cs

```diff
@@ -9,19 +9,34 @@ namespace Microsoft.AspNet.OData.Formatter
 {
     internal class ClrTypeCache
     {
-        private ConcurrentDictionary<Type, IEdmTypeReference> _cache =
+        private ConcurrentDictionary<Type, IEdmTypeReference> _clrToEdmTypeCache =
             new ConcurrentDictionary<Type, IEdmTypeReference>();
 
+        private ConcurrentDictionary<IEdmTypeReference, Type> _edmToClrTypeCache = new ConcurrentDictionary<IEdmTypeReference, Type>();
+
         public IEdmTypeReference GetEdmType(Type clrType, IEdmModel model)
         {
             IEdmTypeReference edmType;
-            if (!_cache.TryGetValue(clrType, out edmType))
+            if (!_clrToEdmTypeCache.TryGetValue(clrType, out edmType))
             {
                 edmType = model.GetEdmTypeReference(clrType);
-                _cache[clrType] = edmType;
+                _clrToEdmTypeCache[clrType] = edmType;
             }
 
             return edmType;
         }
+
+        public Type GetClrType(IEdmTypeReference edmType, IEdmModel edmModel)
```

Comment:
>GetClrType(IEdmTypeReference edmType, IEdmModel edmModel) [](start = 20, length = 57)

We already have the annotation (ClrTypeAnnotation) to hold the EdmType to ClrType,  Is that not enough?

Reply:
This is using that behind the scenes. It called into EdmLibHelpers.GetClrType which uses the annotation. It additionally supports the primitive types as well though and avoid the repeated complex lookup once a type has been discovered.